crates/interpreter/src/gas/calc.rs

```diff
@@ -193,50 +193,56 @@ pub fn sstore_cost<SPEC: Spec>(
     gas: u64,
     is_cold: bool,
 ) -> Option<u64> {
-    // TODO untangle this mess and make it more elegant
+    // EIP-1706 check specific to Istanbul
+    if SPEC::enabled(ISTANBUL) && gas <= CALL_STIPEND {
+        return None;
+    }
+
     let (gas_sload, gas_sstore_reset) = if SPEC::enabled(BERLIN) {
         (WARM_STORAGE_READ_COST, SSTORE_RESET - COLD_SLOAD_COST)
     } else {
         (sload_cost::<SPEC>(is_cold), SSTORE_RESET)
     };
 
-    // https://eips.ethereum.org/EIPS/eip-2200
-    // It’s a combined version of EIP-1283 and EIP-1706
     let gas_cost = if SPEC::enabled(ISTANBUL) {
-        // EIP-1706
-        if gas <= CALL_STIPEND {
-            return None;
-        }
-
-        // EIP-1283
-        if new == current {
-            gas_sload
-        } else {
-            if original == current {
-                if original == U256::ZERO {
-                    SSTORE_SET
-                } else {
-                    gas_sstore_reset
-                }
-            } else {
-                gas_sload
-            }
-        }
+        calculate_istanbul_gas_cost(original, current, new, gas_sload, gas_sstore_reset)
     } else {
-        if current == U256::ZERO && new != U256::ZERO {
-            SSTORE_SET
-        } else {
-            gas_sstore_reset
-        }
+        calculate_non_istanbul_gas_cost(current, new, gas_sstore_reset)
     };
-    // In EIP-2929 we charge extra if the slot has not been used yet in this transaction
+
     if SPEC::enabled(BERLIN) && is_cold {
         Some(gas_cost + COLD_SLOAD_COST)
     } else {
         Some(gas_cost)
     }
 }
 
+fn calculate_istanbul_gas_cost(
```

Comment:
renaming it to `istanbul_sstore_cost` feels better. and put `#[inline(always]` there

Reply:
Resolved this